conn.go

```diff
@@ -58,114 +68,119 @@ func NewTimeoutConn(conn net.Conn, timeout time.Duration) (*Conn, error) {
 func (c *Conn) Close() error {
+	select {
+	case <-c.done:
+		return net.ErrClosed
+	default:
+	}
 	c.Logout()
+	close(c.done)
 	return c.Conn.Close()
 }
 
 // newConn initializes an epp.Conn from a net.Conn.
 // Used internally for testing.
 func newConn(conn net.Conn) *Conn {
-	c := Conn{Conn: conn}
-	c.decoder = xml.NewDecoder(&c.buf)
-	c.saved = *c.decoder
+	c := Conn{
+		Conn:      conn,
+		responses: make(chan *Response),
+		done:      make(chan struct{}),
+	}
+	go c.readLoop()
 	return &c
 }
 
-// reset resets the underlying xml.Decoder and bytes.Buffer,
-// restoring the original state of the underlying
-// xml.Decoder (pos 1, line 1, stack, etc.) using a hack.
-func (c *Conn) reset() {
-	c.buf.Reset()
-	*c.decoder = c.saved // Heh.
-}
-
-// writeDataUnit writes a slice of bytes to c.
-// Bytes written are prefixed with 32-bit header specifying the total size
-// of the data unit (message + 4 byte header), in network (big-endian) order.
-// http://www.ietf.org/rfc/rfc4934.txt
-func (c *Conn) writeDataUnit(x []byte) error {
-	logXML("<-- WRITE DATA UNIT -->", x)
-	s := uint32(4 + len(x))
+// writeRequest writes a single EPP request (x) for writing on c.
+// writeRequest can be called from multiple goroutines.
+func (c *Conn) writeRequest(x []byte) error {
+	c.mWrite.Lock()
+	defer c.mWrite.Unlock()
 	if c.Timeout > 0 {
 		c.Conn.SetWriteDeadline(time.Now().Add(c.Timeout))
 	}
-	err := binary.Write(c.Conn, binary.BigEndian, s)
-	if err != nil {
-		return err
-	}
-	_, err = c.Conn.Write(x)
-	return err
+	return writeDataUnit(c.Conn, x)
 }
 
-// readResponse reads a single EPP response from c and parses the XML into req.
+// readResponse dequeues and returns a EPP response from c.
 // It returns an error if the EPP response contains an error Result.
-func (c *Conn) readResponse(res *Response) error {
-	err := c.readDataUnit()
-	if err != nil {
-		return err
-	}
-	err = IgnoreEOF(scanResponse.Scan(c.decoder, res))
-	if err != nil {
-		return err
-	}
-	if res.Result.IsError() {
-		return &res.Result
+// readResponse can be called from multiple goroutines.
+func (c *Conn) readResponse() (*Response, error) {
+	select {
+	case res := <-c.responses:
+		if res == nil {
+			return res, c.readErr
+		}
+		if res.Result.IsError() {
+			return nil, &res.Result
+		}
+		return res, nil
+	case <-c.done:
+		return nil, net.ErrClosed
 	}
-	return nil
 }
 
-// readDataUnit reads a single EPP message from c into
-// c.buf. The bytes in c.buf are valid until the next
-// call to readDataUnit.
-func (c *Conn) readDataUnit() error {
-	c.reset()
-	var s int32
-	if c.Timeout > 0 {
-		c.Conn.SetReadDeadline(time.Now().Add(c.Timeout))
-	}
-	err := binary.Read(c.Conn, binary.BigEndian, &s)
-	if err != nil {
-		return err
-	}
-	s -= 4 // https://tools.ietf.org/html/rfc5734#section-4
-	if s < 0 {
-		return io.ErrUnexpectedEOF
+func (c *Conn) readLoop() {
+	defer close(c.responses)
+	timeout := c.Timeout
+	r := &io.LimitedReader{R: c.Conn}
```

Comment:
```suggestion
	r := &io.LimitedReader{R: c.Conn, N: 4}
```
Initializing this with `N == 0` likely to be a problem. I'm surprised no test caught this. The docs say
> Read returns EOF when N <= 0

Reply:
Updating `N` below is not a problem.

Reply:
Yep, the hack is to constantly update a single `LimitedReader` instead of creating a new one for each EPP message. It probably saves 1 alloc / cycle, so nbd

Reply:
Pretty sure the read inside `parseDataUnit` will fail unless `N > 0`

Reply:
I now see on line 140 that `parseDataUnit` doesn't get passed `r` rather it is passed `c.Conn`. So the scenario I mentioned above doesn't happen the way it's written here. However, what's here seems like a potential bug, since `r` is wrapping `c.Conn` in a `LimitedReader` and then it reads from `c.Conn` directly before reading through `r`. I'm not sure, it's subtle.

Reply:
An `io.LimitedReader` has only one bit of relevant state, `N`, which is mutated if `Read` is called. Otherwise, reads off the underlying reader have no bearing on the `LimitedReader`.